src/io/linux.zig

```diff
@@ -215,9 +300,128 @@ pub const IO = struct {
+            completion.awaiting_next = first;
+        } else {
+            assert(self.awaiting_count == 0);
+        }
+        self.awaiting = completion;
+        self.awaiting_count += 1;
+
         self.ios_queued += 1;
     }
 
+    /// Cancel should be invoked at most once, before any of the memory owned by read/recv buffers
+    /// is freed (so that lingering async operations do not write to them).
+    ///
+    /// After this function is invoked:
+    /// - No more completion callbacks will be called.
+    /// - No more IO may be submitted.
+    ///
+    /// This function doesn't return until either:
+    /// - All events submitted to io_uring have completed.
+    ///   (They may complete with `error.Canceled`).
+    /// - Or, an io_uring error occurs.
+    ///
+    /// TODO(Linux):
+    /// - Linux kernel ≥5.19 supports the IORING_ASYNC_CANCEL_ALL and IORING_ASYNC_CANCEL_ANY flags,
+    ///   which would allow all events to be cancelled simultaneously with a single "cancel"
+    ///   operation, without IO needing to maintain the `awaiting` doubly-linked list and the `next`
+    ///   cancellation stage.
+    /// - Linux kernel ≥6.0 supports `io_uring_register_sync_cancel` which would remove the `queued`
+    ///   cancellation stage.
+    pub fn cancel(self: *IO) !void {
+        assert(self.cancel_status == .inactive);
+        self.awaiting_verify();
+
+        // Ensure that even if we return early due to an io_uring that IO won't accept more IO.
+        defer self.cancel_status = .done;
+
+        self.cancel_status = .next;
+
+        // Discard any operations that haven't started yet.
+        while (self.unqueued.pop()) |_| {}
+
+        while (self.awaiting) |target| {
+            assert(self.awaiting_count > 0);
+            assert(target.operation != .cancel);
+
+            if (self.cancel_status == .next) {
+                self.cancel_completion = .{
+                    .io = self,
+                    .context = self,
+                    .callback = struct {
+                        fn wrapper(
+                            ctx: ?*anyopaque,
+                            comp: *Completion,
+                            res: *const anyopaque,
+                        ) void {
+                            const io: *IO = @ptrCast(@alignCast(ctx.?));
+                            const result =
+                                @as(*const CancelError!void, @ptrCast(@alignCast(res))).*;
+                            io.cancel_callback(comp, result);
+                        }
+                    }.wrapper,
+                    .operation = .{ .cancel = .{ .target = target } },
+                };
+
+                self.cancel_status = .{ .queued = .{ .target = target } };
+                self.enqueue(&self.cancel_completion);
+            }
+
+            while (self.cancel_status == .queued or self.cancel_status == .wait) {
+                try self.run_for_ns(constants.tick_ms * std.time.ns_per_ms);
```

Comment:
This `try` is suspicious. If we bail out with `try` here, we actually could than get UB? Can we `catch unreachable` here? It seems that, for cancelation, we actually shouldn't be able to error here perhaps? Or maybe not --- there _could_ be some not-yet-submitted IOPS in front of our cancel request, and those can error on submission? 

But perhaps we can guarantee that we just never fail on submission?

Reply:
There are a wide variety of reasons that `run_for_ns` could fail here, including:

```
        // The kernel was unable to allocate memory or ran out of resources for the request.
        // The application should wait for some completions and try again:
        .AGAIN => return error.SystemResources,
        // The SQE `fd` is invalid, or IOSQE_FIXED_FILE was set but no files were registered:
        .BADF => return error.FileDescriptorInvalid,
        // The file descriptor is valid, but the ring is not in the right state.
        // See io_uring_register(2) for how to enable the ring.
        .BADFD => return error.FileDescriptorInBadState,
        // The application attempted to overcommit the number of requests it can have pending.
        // The application should wait for some completions and try again:
        .BUSY => return error.CompletionQueueOvercommitted,
        // The SQE is invalid, or valid but the ring was setup with IORING_SETUP_IOPOLL:
        .INVAL => return error.SubmissionQueueEntryInvalid,
        // The buffer is outside the process' accessible address space, or IORING_OP_READ_FIXED
        // or IORING_OP_WRITE_FIXED was specified but no buffers were registered, or the range
        // described by `addr` and `len` is not within the buffer registered at `buf_index`:
        .FAULT => return error.BufferInvalid,
        .NXIO => return error.RingShuttingDown,
        // The kernel believes our `self.fd` does not refer to an io_uring instance,
        // or the opcode is valid but not supported by this kernel (more likely):
        .OPNOTSUPP => return error.OpcodeNotSupported,
        // The operation was interrupted by a delivery of a signal before it could complete.
        // This can happen while waiting for events with IORING_ENTER_GETEVENTS:
        .INTR => return error.SignalInterrupt,
        else => |errno| return posix.unexpectedErrno(errno),
```

... so the safest thing to do if we hit one is just to crash. That's what the replica does if it hits an error during its normal IO-tick-loop anyhow.